deps/rabbit/src/rabbit_db_exchange.erl

```diff
@@ -0,0 +1,569 @@
+                      false -> fun unconditional_delete_in_mnesia/2
+                  end,
+    rabbit_mnesia:execute_mnesia_transaction(
+      fun() ->
+              case mnesia:wread({?MNESIA_TABLE, XName}) of
+                  [X] -> DeletionFun(X, false);
+                  [] -> {error, not_found}
+              end
+      end).
+
+conditional_delete_in_mnesia(X = #exchange{name = XName}, OnlyDurable) ->
+    case rabbit_db_binding:has_for_source_in_mnesia(XName) of
+        false  -> delete_in_mnesia(X, OnlyDurable, false);
+        true   -> {error, in_use}
+    end.
+
+unconditional_delete_in_mnesia(X, OnlyDurable) ->
+    delete_in_mnesia(X, OnlyDurable, true).
+
+-spec delete_in_mnesia(Exchange, OnlyDurable, RemoveBindingsForSource) -> Ret when
+      Exchange :: rabbit_types:exchange(),
+      OnlyDurable :: boolean(),
+      RemoveBindingsForSource :: boolean(),
+      Exchange :: rabbit_types:exchange(),
+      Binding :: rabbit_types:binding(),
+      Deletions :: dict:dict(),
+      Ret :: {error, not_found} | {error, in_use} | {deleted, Exchange, [Binding], Deletions}.
+delete_in_mnesia(X = #exchange{name = XName}, OnlyDurable, RemoveBindingsForSource) ->
+    ok = mnesia:delete({?MNESIA_TABLE, XName}),
+    mnesia:delete({?MNESIA_DURABLE_TABLE, XName}),
+    rabbit_db_binding:delete_all_for_exchange_in_mnesia(
+      X, OnlyDurable, RemoveBindingsForSource).
+
+%% -------------------------------------------------------------------
+%% delete_serial().
+%% -------------------------------------------------------------------
+
+-spec delete_serial(ExchangeName) -> ok when
+      ExchangeName :: rabbit_exchange:name().
+%% @doc Deletes an exchange serial record from the database.
+%%
+%% @returns ok
+%%
+%% @private
+
+delete_serial(XName) ->
+    rabbit_db:run(
+      #{mnesia => fun() -> delete_serial_in_mnesia(XName) end
+       }).
+
+delete_serial_in_mnesia(XName) ->
+    rabbit_mnesia:execute_mnesia_transaction(
+      fun() ->
+              mnesia:delete({?MNESIA_SERIAL_TABLE, XName})
+      end).
+
+%% -------------------------------------------------------------------
+%% recover().
+%% -------------------------------------------------------------------
+
+-spec recover(VHostName) -> [Exchange] when
+      Exchange :: rabbit_types:exchange(),
+      VHostName :: vhost:name().
+%% @doc Recovers all exchanges for a given vhost
+%%
+%% @returns ok
+%%
+%% @private
+
+recover(VHost) ->
+    rabbit_db:run(
+      #{mnesia => fun() -> recover_in_mnesia(VHost) end
+       }).
+
+recover_in_mnesia(VHost) ->
+    rabbit_mnesia:table_filter(
+      fun (#exchange{name = XName}) ->
+              XName#resource.virtual_host =:= VHost andalso
+                  mnesia:read({?MNESIA_TABLE, XName}) =:= []
+      end,
+      fun (X, true) ->
+              X;
+          (X, false) ->
+              X1 = rabbit_mnesia:execute_mnesia_transaction(
+                     fun() -> set_in_mnesia_tx(X) end),
```

Comment:
I can't find in the original `rabbit_exchange:recover()` where the record is written to the `rabbit_durable_exchange` table. The original exchange code is really difficult to follow.

Do you remember where it is written in `recover()`?

Reply:
Well spotted, it's not. It needs to call just `set_ram_in_mnesia_tx`